op-batcher/batcher/job_runner.go

```diff
@@ -0,0 +1,90 @@
+package batcher
+
+import (
+	"sync"
+)
+
+type JobFactory func() (func(), error)
+
+type JobRunner struct {
+	concurrency uint64
+	started     func(uint64)
+	finished    func(uint64)
+	cond        *sync.Cond
+	wg          sync.WaitGroup
+	running     uint64
+}
+
+// NewJobRunner creates a new JobRunner, with the following parameters:
+//   - concurrency: max number of jobs to run at once (0 == no limit)
+//   - started / finished: called whenever a job starts or finishes. The
+//     number of currently running jobs is passed as a parameter.
+func NewJobRunner(concurrency uint64, started func(uint64), finished func(uint64)) *JobRunner {
+	return &JobRunner{
+		concurrency: concurrency,
+		started:     started,
+		finished:    finished,
+		cond:        sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+// Wait waits on all running jobs to stop.
+func (s *JobRunner) Wait() {
+	s.wg.Wait()
+}
+
+// Run will wait until the number of running jobs is below the max concurrency,
+// and then run the next job. The JobFactory should return `nil` if the next
+// job does not exist. Returns the error returned from the JobFactory (if any).
+func (s *JobRunner) Run(factory JobFactory) error {
```

Comment:
If understand your question, you are asking why this is a "function that, when called, creates the job function", and why can't it just take the job to be run?

We could totally do that. Couple of reasons the implementation is like this:
 - it means the job creation can return an error, which is returned by `Run()`
 - it allows job setup (and the potential for a job to not be available) to happen in the factory

In the batcher's case, when calling `factory()`, it runs the `publishStateToL1Job` setup code, i.e. calling the `l1Tip`,  and calling `state.TxData` to see if there's any transaction data available. If not, the factory will return `nil` indicating that there's no job to run. Otherwise it returns a job `func()` that is run below in the go routine, which performs the transaction send.

Reply:
Removed the factory, because the `l1Tip` being up-to-date doesn't particularly matter if the txData is returned (because the channel was full or the max channel duration was hit if that is the case)